biochatter/api_agent.py

```diff
@@ -0,0 +1,441 @@
+                # to test:
+                # more efficient with \n or without?
+                return "\n".join(lines)
+        except FileNotFoundError:
+            return "The file was not found."
+        except Exception as e:
+            return f"An error occurred: {e}"
+
+    def answer_extraction(self, question: str, file_path: str, n: int) -> str:
+        """
+        Function to extract the answer from the BLAST results.
+
+        Args:
+            question (str): The question to be answered.
+            file_path (str): The path to the BLAST results file.
+            n (int): The number of lines to read from the file.
+
+        Returns:
+            str: The extracted answer from the BLAST results.
+
+        """
+        prompt = ChatPromptTemplate.from_messages(
+            [
+                (
+                    "system",
+                    "You are a world class molecular biologist who knows everything about NCBI and BLAST results.",
+                ),
+                ("user", "{input}"),
+            ]
+        )
+
+        context = self.read_first_n_lines(file_path, n)
+        BLAST_file_answer_extractor_prompt = f"""
+                You have to answer the question: {question} in a clear and concise manner. Be factual!\n\
+                If you are asked what organism a specific sequence belongs to, check the 'Hit_def' fields. If you find a synthetic construct or predicted entry, move to the next one and look for an organism name.\n\
+                Try to use the hits with the best identity score to answer the question. If it is not possible, move to the next one.\n\
+                Be clear, and if organism names are present in ANY of the results, please include them in the answer. Do not make up information and mention how relevant the found information is based on the identity scores.\n\
+                Use the same reasoning for any potential BLAST results. If you find information that is manually curated, please use it and state it. You may also state other results, but always include the context.\n\
+                Based on the information given here:\n\
+                {context}
+                """
+        output_parser = StrOutputParser()
+        chain = prompt | llm | output_parser
+        BLAST_answer = chain.invoke(
+            {"input": {BLAST_file_answer_extractor_prompt}}
+        )
+        return BLAST_answer
+
+
+## Agent class
+class APIAgent:
+    """
+    A class to interact with the BLAST tool for querying and fetching results.
+
+    Attributes:
+        llm (object): The language model to be used.
+        blast_result_path (str): The path to save BLAST results.
+        blast_prompt_path (str): The path to the BLAST prompt file.
+        builder (BlastQueryBuilder): An instance to build BLAST queries.
+        fetcher (BlastFetcher): An instance to fetch BLAST results.
+    """    
+    def __init__(self, llm):
+        self.llm = llm
```

Comment:
In constructors of `RagAgent` and `DatabaseAgent`, we will pass `conversation_factory()` function to create conversation. This is because, in some cases, BioChatter may not have knowledge to create LLM (ChatGPT or Azure ChatGPT).
I think here we can follow the approach of RAGAgent and DatabaseAgent to use conversation_factory() to create conversation, then utilize conversation.chat as llm

Reply:
Absolutely, good observation. Do you have time to suggest the refactor?

Reply:
Yes, I recommend, in APIAgent's constructor, it would be better to pass in conversation_function instead of llm.

Reply:
implemented